src/main/java/com/google/devtools/build/lib/analysis/FilesToRunProvider.java

```diff
@@ -44,6 +44,9 @@ public FilesToRunProvider(
       NestedSet<Artifact> filesToRun,
       @Nullable RunfilesSupport runfilesSupport,
       @Nullable Artifact executable) {
+    if (executable != null) {
+      filesToRun = NestedSetBuilder.fromNestedSet(filesToRun).add(executable).build();
```

Comment:
This will create an extra `NestedSet` instance for every call site, most importantly, in the constructor of `FileConfiguredTarget`, which means an extra object for each input file, which is _probably_ unacceptable overhead.

How about special-casing at least `#fromSingleExecutableArtifact()` so that this overhead is not present?

Given that the only other call site is in `RuleConfiguredTargetBuilder#build()`, I'd prefer putting logic there because, well, that place already has logic. In addition to being a personal preference, you also have more information, notably, `ruleContext.getRule().getRuleClassObject().isSkylark()`, which you can use to determine whether this is a Starlark rule and thus adding the binary to the nested set is necessary.

Even better is changing `SkylarkRuleConfiguredTargetUtil#addSimpleProviders()` -- that place has control over both the `executable and `filesToBuild` fields of `RuleConfiguredTargetBuilder` and is specific to Starlark rules, so you can just add the "executable is forcibly added to `filesToRun`" logic there.

Reply:
`addSimpleProviders()` seems to be used only for constructing `filesToBuild`, but this change is about adding the executable to `filesToRun`.

I've expanded `RuleConfiguredTargetBuilder::build()` (via `buildFilesToRun()`) to inject the executable. It's applied to all rules, since as you note it's difficult to figure out whether an `Artifact` is present in a `NestedSet` without iterating over it.

Reply:
> Given that the only other call site is in `RuleConfiguredTargetBuilder#build()`, I'd prefer putting logic there because, well, that place already has logic.

Done.